### PR TITLE
Support sparseCheckout

### PR DESCRIPTION
**Pull request by BanzaiMan**
This PR resolves merge conflicts in https://github.com/travis-ci/travis-build/pull/882, and addresses the naming concern raised therein.

### CODE REVIEW

File: lib/travis/build/git/clone.rb [hunk 1]
```diff
@@ -22,12 +22,22 @@ def apply
...
-                  sh.echo "Failed to clone from GitHub.", ansi: :red
-                  sh.echo "Checking GitHub status (https://status.github.com/api/last-message.json):"
-                  sh.raw "curl -sL https://status.github.com/api/last-message.json | jq -r .[]"
+              if sparse_checkout
+                sh.cmd "git init #{dir}", assert: true, retry: true
+                sh.cmd "git -C #{dir} config core.sparseCheckout true", assert: true, retry: true
+                sh.cmd "echo #{sparse_checkout} >> #{dir}/.git/info/sparseCheckout", assert: true, retry: true
+                sh.cmd "git -C #{dir} remote add origin #{data.source_url}", assert: true, retry: true
+                sh.cmd "git -C #{dir} pull origin #{branch} #{pull_args}", assert: true, retry: true
+                sh.cmd "cat #{sparse_checkout} >> #{dir}/.git/info/sparseCheckout", assert: true, retry: true
```
Comment (BanzaiMan) [inline lib/travis/build/git/clone.rb]: @PeterDaveHello Is this `cat` correct? I don't think there is a file with this name exists at this point in execution.

**Comment by BanzaiMan** [comment]
@PeterDaveHello I should note here that cdnjs is probably not a very good repository to test this feature. I went through the steps on my machine (assuming that `cat` I mentioned in the comment earlier is really `echo`), and I ended up with a directory with 100 GB of stuff in it. We do not offer VM large enough to hold that much data.

If you would like to move forward, please find a different, smaller repository to test it.

**Comment by BanzaiMan** [comment]
For posterity, here's the build that runs out of disk space on GCE (which has the largest disk capacity we offer): https://staging.travis-ci.org/BanzaiMan/cdnjs/builds/617985

File: lib/travis/build/git/clone.rb (hunk 1 above)
Comment (BanzaiMan) [reply lib/travis/build/git/clone.rb]: Indeed, it fails when the execution actually gets to this spot: https://staging.travis-ci.org/BanzaiMan/travis_staging_test/builds/617995

**Comment by BanzaiMan** [comment]
Further test: https://staging.travis-ci.org/BanzaiMan/travis_staging_test/builds/617997

**Comment by PeterDaveHello** [comment]
@BanzaiMan if the feature works, it won't use that large space :)

**Comment by BanzaiMan** [comment]
@PeterDaveHello Well, that may be true, but I'd much prefer to have a complete end-to-end test done. My test is rather limited and not very indicative of its usefulness.

**Comment by PeterDaveHello** [comment]
Thanks @BanzaiMan !

File: lib/travis/build/git/clone.rb [hunk 2]
```diff
@@ -22,12 +22,20 @@ def apply
...
           def clone_or_fetch
             sh.if "! -d #{dir}/.git" do
-              sh.cmd "git clone #{clone_args} #{data.source_url} #{dir}", assert: true, retry: true
-              if github?
-                sh.if "$? -ne 0" do
-                  sh.echo "Failed to clone from GitHub.", ansi: :red
-                  sh.echo "Checking GitHub status (https://status.github.com/api/last-message.json):"
-                  sh.raw "curl -sL https://status.github.com/api/last-message.json | jq -r .[]"
+              if sparse_checkout
+                sh.echo "Cloning with sparse checkout specificd with #{sparse_checkout}", ansi: :yellow
```
Comment (cotsog) [inline lib/travis/build/git/clone.rb]: Nit: typo `specificd` => `specified`

File: lib/travis/build/git/clone.rb [hunk 3]
```diff
@@ -121,6 +139,15 @@ def source_host_name
 
             host
           end
+
+          def warn_github_status
```
Comment (cotsog) [inline lib/travis/build/git/clone.rb]: Nice!

**Pull request merged**